pkg/plugins/manager/signature/manifestverifier/verifier.go

```diff
@@ -111,20 +112,68 @@ func (pmv *ManifestVerifier) GetPublicKey(keyID string) (string, error) {
 	}()
 
 	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
-		return "", err
+		return err
 	}
 
 	if len(data.Items) == 0 {
-		return "", errors.New("missing public key")
+		return errors.New("missing public key")
 	}
 
 	for _, key := range data.Items {
-		pmv.publicKeys[key.KeyID] = key
+		err = pmv.kv.Set(ctx, key.KeyID, key.PublicKey)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getPublicKey loads public keys from:
+//   - The hard-coded value if the feature flag is not enabled.
+//   - A cached value from kv storage if it has been already retrieved.
+//   - The Grafana.com API if the database is empty.
+func (pmv *ManifestVerifier) GetPublicKey(keyID string) (string, error) {
+	if pmv.cfg == nil || pmv.cfg.Features == nil || !pmv.cfg.Features.IsEnabled("pluginsAPIManifestKey") {
+		return publicKeyText, nil
+	}
+
+	pmv.lock.Lock()
+	defer pmv.lock.Unlock()
+
+	ctx := context.Background()
+	keys, err := pmv.kv.Keys(ctx, "")
+	if err != nil {
+		return "", err
+	}
+	if len(keys) == 0 {
+		// Populate with the default key
+		err := pmv.kv.Set(ctx, keyID, publicKeyText)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	key, exist, err := pmv.kv.Get(ctx, keyID)
+	if err != nil {
+		return "", err
+	}
+	if exist {
+		return key, nil
+	}
+
+	// Not in the cache, download keys
```

Comment:
it may be worth keeping track of the last time the keys were downloaded so that a plugin with a bad signature doesn't constantly trigger re-downloading the keys.

Reply:
Removed this synchronous download, now it always happen in parallel.